README.md
<!-- 
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages). 

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages). 
-->

Responsive Side Menu Scaffold

## Features



## Getting started

Replace normal Scaffold with This and add Menu.

## Usage



```dart
ResponsiveScaffold(
                key: context.read<SideBarCubit>().scaffoldKey,
                body: const AutoRouter(),
                menuTitle: Text(S.of(context).DcartAdminPanel),
                // Make Rail width larger when using it on tablet or desktop.
                railWidth: 66,
                appBar: showNab(context)
                    ? AppBar(
                        backgroundColor: Theme.of(context).cardColor,
                        elevation: 0,
                        iconTheme: Theme.of(context).iconTheme,
                        actions: [
                          IconButton(
                            onPressed: () {},
                            icon: const Icon(
                              Icons.notifications_outlined,
                            ),
                          ),
                          const SizedBox(
                            width: 10,
                          ),
                        ],
                      )
                    : null,
                menuHeader: const MenuHeader(),
                menuItems: [
                  MenuItemInfo(
                      icon: Icons.person_outline, label: S.of(context).profile),
                  MenuItemInfo(
                      icon: Icons.dashboard_outlined,
                      label: S.of(context).dashboard),
                  MenuItemInfo(
                      icon: Icons.people_alt_outlined,
                      label: S.of(context).employees),
                  MenuItemInfo(icon: Icons.tab, label: S.of(context).tables),
                  MenuItemInfo(
                      icon: Icons.restaurant_menu_rounded,
                      label: S.of(context).menu),
                  MenuItemInfo(
                      icon: Icons.shopping_cart,
                      label: S.of(context).eventItems),
                  MenuItemInfo(
                      icon: Icons.storefront,
                      label: S.of(context).businessServices),
                  MenuItemInfo(
                      icon: Icons.park_outlined,
                      label: S.of(context).dailyActivities),
                ],
                currentSelectedMenuItem: _determineIndexByRoute(
                    AutoRouter.of(context, watch: true)
                            .innerRouterOf(HomePageRoute.name)
                            ?.current
                            .name ??
                        ""),
                onSelect: (int index) {
                  switch (index) {
                    case 0:
                      _onPressedItem(context, const ProfilePageRoute());
                      break;
                    case 1:
                      _onPressedItem(context, const DashboardRoute());
                      break;
                    case 2:
                      _onPressedItem(context, const EmployeeListPageRoute());
                      break;
                    case 3:
                      _onPressedItem(
                          context, const ServingLocationManagementPageRoute());
                      break;
                    case 4:
                      _onPressedItem(context, const MenuPageRoute());
                      break;
                    case 5:
                      _onPressedItem(context, const EventItemsPageRoute());
                      break;
                    case 6:
                      _onPressedItem(context, BusinessServicesPageRoute());
                      break;
                    case 7:
                      _onPressedItem(context, const DailyActivitiesPageRoute());
                      break;

                    default:
                  }
                },
              ),
```

## Additional information



CHANGELOG.md
## 0.0.4

* Upgrade Dep
## 0.0.3

* Breaking Change Api for selected Item in Menu using an Id.
## 0.0.2

* Little Fix
## 0.0.1

* First Release.



example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:responsive_sidebar_example/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(const MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


test/responsive_sidebar_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:responsive_sidebar/responsive_sidebar.dart';

void main() {
  test('adds one to input values', () {});
}


example/windows/runner/main.cpp
#include <flutter/dart_project.h>
#include <flutter/flutter_view_controller.h>
#include <windows.h>

#include "flutter_window.h"
#include "utils.h"

int APIENTRY wWinMain(_In_ HINSTANCE instance, _In_opt_ HINSTANCE prev,
                      _In_ wchar_t *command_line, _In_ int show_command) {
  // Attach to console when present (e.g., 'flutter run') or create a
  // new console when running with a debugger.
  if (!::AttachConsole(ATTACH_PARENT_PROCESS) && ::IsDebuggerPresent()) {
    CreateAndAttachConsole();
  }

  // Initialize COM, so that it is available for use in the library and/or
  // plugins.
  ::CoInitializeEx(nullptr, COINIT_APARTMENTTHREADED);

  flutter::DartProject project(L"data");

  std::vector<std::string> command_line_arguments =
      GetCommandLineArguments();

  project.set_dart_entrypoint_arguments(std::move(command_line_arguments));

  FlutterWindow window(project);
  Win32Window::Point origin(10, 10);
  Win32Window::Size size(1280, 720);
  if (!window.CreateAndShow(L"responsive_sidebar_example", origin, size)) {
    return EXIT_FAILURE;
  }
  window.SetQuitOnClose(true);

  ::MSG msg;
  while (::GetMessage(&msg, nullptr, 0, 0)) {
    ::TranslateMessage(&msg);
    ::DispatchMessage(&msg);
  }

  ::CoUninitialize();
  return EXIT_SUCCESS;
}


lib/src/components/app_menu.dart
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:responsive_sidebar/src/components/shadow_card.dart';

import 'menu_item.dart';

/// App side menu and rail

class AppMenu extends StatefulWidget {
  const AppMenu({
    Key? key,
    this.title,
    required this.maxWidth,
    this.onOperate,
    this.onSelect,
    required this.railWidth,
    required this.menuItems,
    this.menuFooter,
    this.menuHeader,
    this.selectedItemId,
  }) : super(key: key);
  final Widget? title;
  final double maxWidth;
  final VoidCallback? onOperate;
  //When Menu Item Is Selected
  final ValueChanged<String>? onSelect;
  final double railWidth;
  final List<MenuItemInfo> menuItems;
  final String? selectedItemId;

  ///Widget rendered before the [SideMenuItem] list
  final Widget? menuHeader;

  ///Widget rendered after the [SideMenuItem] list
  final Widget? menuFooter;

  @override
  _AppMenuState createState() => _AppMenuState();
}

class _AppMenuState extends State<AppMenu> {
  String? selectedItemId;

  @override
  void didUpdateWidget(covariant AppMenu oldWidget) {
    if (widget.selectedItemId != oldWidget.selectedItemId) {
      selectedItemId = widget.selectedItemId;
    }
    super.didUpdateWidget(oldWidget);
  }

  @override
  void initState() {
    selectedItemId = widget.selectedItemId;
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return LayoutBuilder(
      builder: (BuildContext context, BoxConstraints size) {
        // The overflow box is not the prettiest approach, but has some
        // convenient properties for this simple animated menu using just
        // a single AnimatedContainer.
        return OverflowBox(
          alignment: AlignmentDirectional.topStart,
          minWidth: 0,
          maxWidth: widget.maxWidth,
          child: ShadowCard(
            width: size.maxWidth,
            borderRadius: const BorderRadius.all(Radius.zero),
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: <Widget>[
                // We use an AppBar element as header in the menu too, a custom
                // Widget would be less restrictive, but for simplicity, the
                // AppBar has so many nice things built in to handle text style,
                // size and scaling for the title that are tedious to replicate
                Padding(
                  padding: kIsWeb
                      ? const EdgeInsets.all(0)
                      : const EdgeInsets.only(top: 30.0),
                  child: Row(
                    mainAxisAlignment: MainAxisAlignment.start,
                    children: [
                      Flexible(
                        child: Padding(
                          padding: const EdgeInsets.all(14.0),
                          child: IconButton(
                            onPressed: () {
                              widget.onOperate?.call();
                            },
                            tooltip: "Menu",
                            icon: Icon(
                              Icons.menu_open,
                              color: Theme.of(context).iconTheme.color,
                            ),
                          ),
                        ),
                      ),
                      if (size.maxWidth > 150)
                        Flexible(child: widget.title ?? const SizedBox())
                    ],
                  ),
                ),
                if (widget.menuHeader != null) widget.menuHeader!,
                Expanded(
                  child: ListView(
                    padding: EdgeInsets.zero, //  Removes all edge insets
                    children: <Widget>[
                      // A mock user profile on the menu/rail.
                      // _UserProfile(railWidth: widget.railWidth),
                      // Add all the menu items.

                      for (int i = 0; i < widget.menuItems.length; i++)
                        SideMenuItem(
                          width: size.maxWidth,
                          menuWidth: widget.maxWidth,
                          onTap: () {
                            setState(() {
                              selectedItemId = widget.menuItems[i].id;
                            });
                            widget.onSelect?.call(widget.menuItems[i].id);
                          },
                          selected: selectedItemId == widget.menuItems[i].id,
                          icon: widget.menuItems[i].icon,
                          label: widget.menuItems[i].label,
                          showDivider: false,
                          railWidth: widget.railWidth,
                        ),
                    ],
                  ),
                ),
                if (widget.menuFooter != null) widget.menuFooter!
              ],
            ),
          ),
        );
      },
    );
  }
}


lib/src/components/menu_item.dart
import 'package:flutter/material.dart';

import 'maybe_tooltip.dart';

/// The items for the menu.
class SideMenuItem extends StatelessWidget {
  const SideMenuItem({
    Key? key,
    required this.width,
    required this.menuWidth,
    required this.onTap,
    this.selected = false,
    required this.icon,
    required this.label,
    this.showDivider = false,
    required this.railWidth,
  }) : super(key: key);

  final double width;
  final double menuWidth;
  final VoidCallback onTap;
  final bool selected;
  final IconData icon;
  final String label;
  final bool showDivider;
  final double railWidth;

  // Height of the menu item.
  static const double _itemHeight = 42;

  @override
  Widget build(BuildContext context) {
    final ThemeData theme = Theme.of(context);
    final bool isLight = theme.brightness == Brightness.light;
    // Just some colors for the menu items, based on the current color schemes.
    // Always when you can, use colors from the theme.colorScheme to make
    // custom elements in your app they react to theme changes and use the theme
    // colors. You can make elaborate hues and opacities of the colors in the
    // theme's color schemes, like here:
    final Color iconColor = isLight
        ? Color.alphaBlend(theme.colorScheme.primary.withAlpha(0x99),
            theme.colorScheme.onBackground)
        : Color.alphaBlend(theme.colorScheme.primary.withAlpha(0x99),
            theme.colorScheme.onBackground);

    final Color textColor =
        theme.textTheme.bodyLarge?.color ?? theme.colorScheme.primary;

    // The M3 guide calls for 12dp padding after the selection indicator on
    // the menu highlight in a Drawer or side menu. We can do that, but we
    // have such a narrow rail for phone size, so at rail sizes we will make it
    // much smaller, even 2 different sizes.
    final double endPadding = (width > railWidth + 10)
        ? 12
        // If we use a really narrow rail rail, make padding even smaller-
        : railWidth < 60
            ? 5
            : 8;
    // Remove the menu when it gets smaller than 4dp during animation.
    if (width < 4) {
      return const SizedBox.shrink();
    } else {
      return Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: <Widget>[
          if (showDivider) const Divider(thickness: 1, height: 1),
          Padding(
            padding:
                EdgeInsetsDirectional.fromSTEB(endPadding, 2, endPadding, 2),
            child: Material(
              clipBehavior: Clip.antiAlias,
              borderRadius: const BorderRadius.all(Radius.circular(5)),
              // This is a tap command menu, so we keep selected item
              // transparent. We still get desktop/web hover and tap splash.
              // However, you can do this on a menu that selects an item that
              // should remain highlighted and show last selected item:
              color: selected
                  ? theme.primaryColor.withAlpha(50)
                  : Colors.transparent,
              //color: Colors.transparent,
              child: InkWell(
                onTap: onTap,
                child: SizedBox(
                  height: _itemHeight,
                  width: width,
                  child: OverflowBox(
                    alignment: AlignmentDirectional.topStart,
                    minWidth: 0,
                    maxWidth: menuWidth,
                    child: Row(
                      children: <Widget>[
                        MaybeTooltip(
                          // Show tooltips only at rail size.
                          condition: width == railWidth,
                          // The item menu labels is a tooltip on rail size.
                          message: label,
                          // Just to get the tooltip outside the rail.
                          margin: const EdgeInsetsDirectional.only(start: 50),
                          // Constrain icon to min of rail width.
                          child: Padding(
                            padding: const EdgeInsets.symmetric(
                                vertical: 8.0, horizontal: 14),
                            child: Icon(icon,
                                color: selected
                                    ? iconColor
                                    : textColor.withOpacity(0.8)),
                          ),
                        ),
                        // Below width of 10dp we remove the label.
                        if (width < railWidth + 10)
                          const SizedBox.shrink()
                        else
                          Text(
                            label,
                            style: theme.textTheme.bodyLarge?.copyWith(
                                color: selected
                                    ? iconColor
                                    : textColor.withOpacity(0.8)),
                          )
                      ],
                    ),
                  ),
                ),
              ),
            ),
          ),
        ],
      );
    }
  }
}

class MenuItemInfo {
  final IconData icon;
  final String label;

  /// Unique Id
  final String id;

  MenuItemInfo({
    required this.id,
    required this.icon,
    required this.label,
  });
}


lib/src/components/shadow_card.dart
import 'package:flutter/material.dart';

class ShadowCard extends StatelessWidget {
  const ShadowCard({
    Key? key,
    this.child,
    this.backgroundColor,
    this.gradient,
    this.border,
    this.borderRadius,
    this.height,
    this.width,
    this.margin,
    this.elevetion,
    this.constraints,
  }) : super(key: key);
  final Widget? child;
  final Color? backgroundColor;
  final LinearGradient? gradient;
  final Border? border;
  final BorderRadius? borderRadius;
  final double? height;
  final double? width;
  final EdgeInsetsGeometry? margin;
  final double? elevetion;
  final BoxConstraints? constraints;
  @override
  Widget build(BuildContext context) {
    return Container(
      height: height,
      width: width,
      margin: margin ?? const EdgeInsets.all(0),
      constraints: constraints,
      decoration: BoxDecoration(
          gradient: gradient,
          color: backgroundColor ?? Theme.of(context).cardColor,
          borderRadius:
              borderRadius ?? const BorderRadius.all(Radius.circular(10)),
          border: border,
          boxShadow: <BoxShadow>[
            if (Theme.of(context).brightness == Brightness.light)
              BoxShadow(
                blurRadius: 8,
                color: Theme.of(context).primaryColor.withAlpha(20),
              )
          ]),
      child: ClipRRect(
        borderRadius:
            borderRadius ?? const BorderRadius.all(Radius.circular(10)),
        child: Material(
          color: Colors.transparent,
          elevation: elevetion ?? 0.0,
          child: child,
        ),
      ),
    );
  }

  factory ShadowCard.bordered({
    required final BuildContext context,
    required final Widget child,
    final EdgeInsetsGeometry? margin,
    final double? height,
    final double? width,
    final BorderRadius? borderRadius,
  }) {
    return ShadowCard(
      margin: margin,
      height: height,
      width: width,
      borderRadius: borderRadius,
      backgroundColor: Theme.of(context).colorScheme.secondary,
      border: Theme.of(context).brightness == Brightness.light
          ? Border.all(color: Theme.of(context).primaryColor.withAlpha(20))
          : null,
      child: child,
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:responsive_sidebar/responsive_sidebar.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      themeMode: ThemeMode.dark,
      theme: ThemeData(
        primarySwatch: Colors.cyan,
      ),
      darkTheme:
          ThemeData(primarySwatch: Colors.cyan, brightness: Brightness.light),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    return ResponsiveScaffold(
      appBar: AppBar(
        title: const Text("Example Sidebar Scaffold"),
        actions: const <Widget>[],
        elevation: 0,
        // Some logic to show the implicit menu button on AppBar when
        // there is no rail or menu.

        // !isDesktop && isMenuClosed && menuDoneClosing,
      ),
      menuTitle: const Text("Example Sidebar Scaffold"),
      // Make Rail width larger when using it on tablet or desktop.
      railWidth: 66,
      breakpointShowFullMenu: 900,
      extendBodyBehindAppBar: false,
      extendBody: true,
      menuItems: [
        MenuItemInfo(
          id: "Dashboard",
          icon: Icons.dashboard,
          label: "Dashboard",
        ),
        MenuItemInfo(id: "Pacientes", icon: Icons.people, label: "Pacientes"),
        MenuItemInfo(
            id: "Estudios", icon: Icons.medical_services, label: "Estudios"),
      ],
      menuHeader: const _UserProfile(railWidth: 52),
      menuFooter: const Padding(
        padding: EdgeInsets.all(8.0),
        child: Text("Version 1.0"),
      ),
      onSelect: (id) {
        debugPrint(id);
      },
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            const Text(
              'You have pushed the button this many times:',
            ),
            Text(
              '$_counter',
              style: Theme.of(context).textTheme.headlineMedium,
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: const Icon(Icons.add),
      ),
    );
  }
}

/// A user profile widget that we use as leading widget in the menu.
///
/// Mock UI for the demo app, it does not do anything.
class _UserProfile extends StatefulWidget {
  const _UserProfile({
    Key? key,
    required this.railWidth,
  }) : super(key: key);
  final double railWidth;

  @override
  _UserProfileState createState() => _UserProfileState();
}

class _UserProfileState extends State<_UserProfile> {
  bool _collapsed = true;

  @override
  Widget build(BuildContext context) {
    final ThemeData theme = Theme.of(context);
    final TextTheme textTheme = theme.textTheme;
    final TextTheme primaryTextTheme = theme.primaryTextTheme;
    const double hPadding = 5;
    return LayoutBuilder(
      builder: (BuildContext context, BoxConstraints constraints) {
        print(constraints.maxWidth);
        return Material(
          // As an effect for theme demos, we put themed surface color on
          // Material used as background for the user profile widget. This gives
          // a it a slightly different tone for the background with themes that use
          // a blend mode where the blend strength is different for background
          // and surface, since this widget is placed on Material with background
          // color it will show up in such theme modes.
          color: theme.colorScheme.surface,
          child: Column(
            children: <Widget>[
              Row(
                children: [
                  Padding(
                    padding: const EdgeInsets.all(10.0),
                    child: CircleAvatar(
                      backgroundColor: theme.colorScheme.primary,
                      radius: widget.railWidth / 2 - hPadding,
                      child: Text(
                        'MR',
                        style: primaryTextTheme.titleMedium!.copyWith(
                            color: theme.colorScheme.onPrimary,
                            fontWeight: FontWeight.w600),
                      ),
                    ),
                  ),
                  if (constraints.maxWidth > widget.railWidth * 2) ...[
                    Expanded(
                      child: Text(
                        'Marcos Rodriguez',
                        maxLines: 1,
                        style: textTheme.titleMedium!
                            .copyWith(fontWeight: FontWeight.w600),
                      ),
                    ),
                    ExpandIcon(
                      isExpanded: !_collapsed,
                      size: 24,
                      padding: EdgeInsets.zero,
                      onPressed: (_) {
                        setState(() {
                          _collapsed = !_collapsed;
                        });
                      },
                    ),
                  ]
                ],
              ),
              // Add some expand actions for access to mock functionality.
              AnimatedSwitcher(
                duration: const Duration(milliseconds: 200),
                transitionBuilder: (Widget child, Animation<double> animation) {
                  return SizeTransition(
                    sizeFactor: animation,
                    child: child,
                  );
                },
                child: _collapsed
                    ? const SizedBox.shrink()
                    : Padding(
                        padding: const EdgeInsets.only(bottom: 8),
                        child: Row(
                          children: <Widget>[
                            const Spacer(),
                            TextButton(
                              onPressed: () {},
                              child: Column(
                                children: <Widget>[
                                  const Icon(Icons.person),
                                  Text('Profile', style: textTheme.labelSmall),
                                ],
                              ),
                            ),
                            const SizedBox(width: 10),
                            TextButton(
                              onPressed: () {},
                              child: Column(
                                children: <Widget>[
                                  const Icon(Icons.logout),
                                  Text('Sign out', style: textTheme.labelSmall),
                                ],
                              ),
                            ),
                            const SizedBox(width: 8),
                          ],
                        ),
                      ),
              ),
            ],
          ),
        );
      },
    );
  }
}
